prediction_market_agent/agents/blockchain_coding_agent/functions.py

```diff
@@ -0,0 +1,143 @@
+import typing as t
+
+import requests
+import tenacity
+from autogen import ConversableAgent, register_function
+from tavily import TavilyClient
+
+from prediction_market_agent.utils import APIKeys
+
+
+@tenacity.retry(stop=tenacity.stop_after_attempt(3), wait=tenacity.wait_fixed(1))
+def tavily_search(
+    query: str,
+) -> dict[str, t.Any]:
+    """
+    Internal minimalistic wrapper around Tavily's search method, that will retry if the call fails.
+    """
+    tavily = TavilyClient(api_key=(APIKeys()).tavily_api_key.get_secret_value())
+    response: dict[str, t.Any] = tavily.search(query=query)
+    return response
+
+
+from typing import List
+
+from eth_typing import ChecksumAddress
+from prediction_market_agent_tooling.markets.omen.omen_contracts import (
+    OmenConditionalTokenContract,
+)
+from prediction_market_agent_tooling.tools.contract import ContractOnGnosisChain
+from web3 import Web3
+
+
+def fetch_read_methods_from_blockscout(contract_address: str) -> t.Any:
+    w3 = OmenConditionalTokenContract().get_web3()
+    if not is_contract(w3, Web3.to_checksum_address(contract_address)):
+        raise ValueError(f"{contract_address=} is not a contract on Gnosis Chain.")
+    read_not_proxy = fetch_read_methods(contract_address)
+    read_proxy = fetch_read_methods_proxy(contract_address)
+    # ToDo - Fetch write methods
+    return read_not_proxy + read_proxy
+
+
+def fetch_read_methods(contract_address: str) -> t.Any:
+    url = f"https://gnosis.blockscout.com/api/v2/smart-contracts/{contract_address}/methods-read?is_custom_abi=false"
+    r = requests.get(url)
+    return r.json()
+
+
+def is_contract(web3: Web3, contract_address: ChecksumAddress) -> bool:
+    return bool(web3.eth.get_code(contract_address))
+
+
+def fetch_read_methods_proxy(contract_address: str) -> t.Any:
+    url = f"https://gnosis.blockscout.com/api/v2/smart-contracts/{contract_address}/methods-read-proxy?is_custom_abi=false"
+    r = requests.get(url)
+    return r.json()
+
+
+def get_rpc_endpoint() -> str:
+    return "https://rpc.gnosischain.com"
+
+
+def checksum_address(address: str) -> ChecksumAddress:
+    from web3 import Web3
+
+    return Web3.to_checksum_address(address)
```

Comment:
Is there some reason to wrap a single call to `Web3.to_checksum_address` with a single call to checksum_address? 🤔

Reply:
Oh I guess the function needs to be like this, because you can register Web3.to_checksum as agent's function?